### PR TITLE
🎉 Source Airtable: implement OAuth 

### LINKED ISSUE #20708: 🎉 Source Airtable: Implement OAuth2.0 authentication
Based on the official docs: https://airtable.com/developers/web/api/oauth-reference
the Oauth flow is also supported and could be implemented to access Airtable resources.

### PR DESCRIPTION
**Pull request by bazarnov**
## What
Resolving: https://github.com/airbytehq/airbyte/issues/20708

## How
- updated the connector to work with `SingleUseRefreshTokens`
- added Java part for server
- tested on Local Dev UI
- added  the `prod` OAuth2.0 credentials to the `prod-db`

## 🚨 User Impact 🚨
No impact is expected if OLD credentials are used already.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

### CODE REVIEW

**Comment by bazarnov** [comment]
Blocked by: https://github.com/airbytehq/airbyte/issues/20145

@pedroslopez Please consider this one as dependency, we need to cover `check` and `discover` stages with `ConfigUpdate` to have the oAuth for Airtable working.

File: airbyte-integrations/connectors/source-airtable/source_airtable/auth.py
Comment (antixar) [inline airbyte-integrations/connectors/source-airtable/source_airtable/auth.py:51]: Hi, 
Please pay your attention to the next points:
1. Any `__new__` magic function accepts an instance as the first argument. And it will be better if it is "cls".
2. `config["api_key"]`  can be empty or `config` can be `None`. I would have check by this if had written this condition: `(config or {}).get("api_key")`.

File: airbyte-integrations/connectors/source-airtable/source_airtable/source.py
Comment (antixar) [inline airbyte-integrations/connectors/source-airtable/source_airtable/source.py:45]: You can use more simple and readable variant:
`auth = self._auth or AirtableAuth(config)`

File: airbyte-integrations/connectors/source-airtable/source_airtable/auth.py
Comment (antixar) [inline airbyte-integrations/connectors/source-airtable/source_airtable/auth.py:26]: Try to use ready functions: [get_client_id](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py#L184) and [get_client_secret](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py#L187).

File: airbyte-integrations/connectors/source-airtable/source_airtable/source.py
Comment (antixar) [inline airbyte-integrations/connectors/source-airtable/source_airtable/source.py:23]: I see to the class variable `_auth` is used by 2 diff functions: `streams` and `check_connection`. I am not sure to these ones can be called by same python script call and this code optimisation will have some benefit because AirByte engine calls them by diff commands. Please correct me if I am not right.

File: airbyte-integrations/connectors/source-airtable/source_airtable/spec.json
Comment (antixar) [inline airbyte-integrations/connectors/source-airtable/source_airtable/spec.json:51]: if we know its permanent length or some RegExp, we can set this additional validation here.

Comment (bazarnov) [reply airbyte-integrations/connectors/source-airtable/source_airtable/spec.json:51]: No need to RegExp of alphanumeric chars, especially if this could be changed on the API side, this would require the spec tunings later on, which is not necessary I think.

File: airbyte-integrations/connectors/source-airtable/source_airtable/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-airtable/source_airtable/source.py:23]: You're correct, therefore in order to have a local test and development more manageable, we need to save the auth for later use, if possible. The Platform calls the check, discover, and streams using different commands; therefore, while using `SingleUseRefreshToken` we need to be sure that the same `access_token` acquired from OAuth2.0 will be passed to all the stream class instances and will not be triggered for each stream generated.

File: airbyte-integrations/connectors/source-airtable/source_airtable/auth.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-airtable/source_airtable/auth.py:26]: Perfectly fitted)

**Review by lazebnyi — APPROVED** [review]

File: airbyte-integrations/connectors/source-airtable/acceptance-test-config.yml
Comment (alafanechere) [inline airbyte-integrations/connectors/source-airtable/acceptance-test-config.yml:24]: Is this the same bypass reason as for 0.1.3?

File: airbyte-integrations/connectors/source-airtable/integration_tests/invalid_config_oauth.json
Comment (alafanechere) [inline airbyte-integrations/connectors/source-airtable/integration_tests/invalid_config_oauth.json:10]: missing end line

File: airbyte-integrations/connectors/source-airtable/source_airtable/source.py
Comment (alafanechere) [inline airbyte-integrations/connectors/source-airtable/source_airtable/source.py:45]: Could you please explain this logic?

File: airbyte-integrations/connectors/source-airtable/source_airtable/spec.json
Comment (alafanechere) [inline airbyte-integrations/connectors/source-airtable/source_airtable/spec.json:7]:
```suggestion
```
True is the default.

**Review by alafanechere — APPROVED** [review]
Code LGTM but I think we should pin the airtable version in `airbyte-cloud` to make sure the java code is deployed before we update the connector version.

Please also sync with @airbytehq/cloud-support before merging / deploying the new connector version on cloud as there is a breaking change related to configuration. Could you explain what actions the users need to take to make their sync working again?

cc. @pedroslopez are you comfortable having a second "single use refresh token" connector on cloud or shall we hold until your current platform improvement are live?

File: airbyte-integrations/connectors/source-airtable/integration_tests/expected_records.txt
Comment (alafanechere) [inline airbyte-integrations/connectors/source-airtable/integration_tests/expected_records.txt:1]: please use the `.jsonl` extension instead of `.txt`

Comment (bazarnov) [reply airbyte-integrations/connectors/source-airtable/integration_tests/expected_records.txt:1]: Tricky part) It used to be `txt`  as far as I remember [here](https://docs.airbyte.com/connector-development/testing-connectors/source-acceptance-tests-reference/#example-of-expected_recordsjsonl). 

Sure, thanks.

File: airbyte-integrations/connectors/source-airtable/acceptance-test-config.yml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-airtable/acceptance-test-config.yml:24]: This part could be removed for this particular version release, there are no changes with the previous version, so it should be backward capable with the previous version.

**Comment by erica-airbyte** [comment]
@alafanechere or @bazarnov looks like this has been deployed?

From @alafanechere note it sounds like this is a breaking change? 

If so please provide the below information
- Workspaces affected by the change
- Time when the change will be applied
- Reason for the breaking change
- What is being updated/changed
- What will the customer see that is different? 
- What actions the customers need to take once the changes goes live? (Refresh source schema, Select to reset affected streams, Reset, Update permissions (if this why are permissions changing))

File: airbyte-integrations/connectors/source-airtable/source_airtable/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-airtable/source_airtable/source.py:45]: We need to trigger the catalog generation to produce streams from the auto-generated catalog and not generate it 2-ce, take a look at the `streams` method, it reveals this mystery)

File: airbyte-integrations/connectors/source-airtable/source_airtable/spec.json
Comment (bazarnov) [reply airbyte-integrations/connectors/source-airtable/source_airtable/spec.json:7]: It should be `True` to provide the backward capability.

**Comment by bazarnov** [comment]
> Please also sync with @airbytehq/cloud-support before merging / deploying the new connector version on cloud as there is a breaking change related to configuration. Could you explain what actions the users need to take to make their sync working again?

@alafanechere 
No breaking changes are expected with this PR, it doesn't touch the Customers that are already has their connections. The old connections will still work - tested by setting up the old version > updated to this one > run the sync again >  looks good.


> @alafanechere or @bazarnov looks like this has been deployed?
> 
> From @alafanechere note it sounds like this is a breaking change?
> 
> If so please provide the below information
> 
> * Workspaces affected by the change
> * Time when the change will be applied
> * Reason for the breaking change
> * What is being updated/changed
> * What will the customer see that is different?
> * What actions the customers need to take once the changes goes live? (Refresh source schema, Select to reset affected streams, Reset, Update permissions (if this why are permissions changing))

@erica-airbyte No at all, these changes are not  deployed yet) we mention your team here to notify that this is going to be deployed soon)

**Comment by pedroslopez** [comment]
> cc. @pedroslopez are you comfortable having a second "single use refresh token" connector on cloud or shall we hold until your current platform improvement are live?

All good from my end!

**Comment by bazarnov** [comment]
> Agree we should do the following ahead of time though:

Sure.

**Comment by bazarnov** [comment]
@erica-airbyte 

> * Workspaces affected by the change

Hard to say close enough but maybe 2-5

> * Time when the change will be applied

After [This Cloud PR](https://github.com/airbytehq/airbyte-cloud/pull/4212) is reverted back and version is >=1.0.1

> * Reason for the breaking change (is it breaking?)

Users must have OAuth for `source-airtable`, it simplifies the setting right scopes and selecting the other user-related sections for authorization purposes, which are important for the schema auto-generation mechanisms to provide the dynamic schema discovery based on the authenticated user in this PR https://github.com/airbytehq/airbyte/pull/21851 
However it's not directly implied the possibility to generate schemas automatically, OAuth gives it a better shine.

> * What will the customer see that is different?

The customer will see 2 options instead of 1 on the set-up source page, for `source-airtable`. 

> * What actions the customers need to take once the changes goes live?
No actions are required.

**Comment by bazarnov** [comment]
Version `1.0.1` is pinned for Cloud, after the server components are updated and released to Cloud, the PR: https://github.com/airbytehq/airbyte-cloud/pull/4212 should be reverted and the version should be unpinned to catch the latest image.

**Pull request merged**
